### PR TITLE
docs(utils): properly document is_normalized_name vs canonicalize_name

### LINKED ISSUE #1231: is_normalized_name() rejects some names that canonicalize_name() leaves unchanged (leading/trailing hyphens)
Spotted by @henryiii while reviewing #1230.

`is_normalized_name(name)` is documented to return `True` iff the name is already canonical (`canonicalize_name(name) == name`, i.e. it roundtrips). That holds in the interior but breaks at the edges: `canonicalize_name` can emit names with leading/trailing hyphens (it lowercases and collapses `[-_.]+` to `-` without stripping edge hyphens), yet `is_normalized_name` rejects those.

Reproducer (packaging 26.0):

```python
from packaging.utils import canonicalize_name, is_normalized_name
n = canonicalize_name("_not_legal")   # -> "-not-legal"
print(canonicalize_name(n) == n)       # True  (already canonical / roundtrips)
print(is_normalized_name(n))           # False (contract says this should be True)
```

Same for `"a-"`, `"-a"`, `"-not-legal"`. An exhaustive `{a,1,-}`-style enumeration shows the divergence is exactly the leading/trailing-hyphen class (henryiii counted 895 such names in the #1230 review).

Root cause: `is_normalized_name` matches `[a-z0-9]+(?:-[a-z0-9]+)*` (PEP 503 valid-name shape — no leading/trailing hyphen), while `canonicalize_name` never strips edge hyphens, so it can produce strings the regex won't accept.

These aren't valid distribution names, so the practical impact is narrow, but the two functions disagree on their documented relationship. Resolution is a maintainer call — tighten `canonicalize_name` to strip edge hyphens, relax `is_normalized_name` to mirror `canonicalize_name`'s output, or document the edge in the "iff roundtrips" wording. Filing per @henryiii's note in #1230 to keep track.

### PR DESCRIPTION
**Pull request by gaoflow**
## Summary

`is_normalized_name()` is documented to return `True` iff `canonicalize_name()` would roundtrip to the same value. However, the previous regex-based implementation (`[a-z0-9]+(?:-[a-z0-9]+)*`) rejected names with leading or trailing hyphens — exactly the strings `canonicalize_name()` can produce for inputs like `"_not_legal"` or `"a_"`.

```python
>>> from packaging.utils import canonicalize_name, is_normalized_name
>>> n = canonicalize_name("_not_legal")   # -> "-not-legal"
>>> canonicalize_name(n) == n             # True (fixed point)
True
>>> is_normalized_name(n)                 # False — breaks the documented contract
False
```

The fix replaces the regex check with the definitive roundtrip check the docs describe.

Fixes #1231.

---

This pull request was prepared with the assistance of AI, under my direction and review.

### CODE REVIEW

**Comment by henryiii** [comment]
Won't this have a performance impact? Checking something is cheaper than modifying a string, I'd expect? I'll trigger the CI and see if we have a perf test for this.

**Comment by binggao1230** [comment]
Pushed `eab1122` to address the lint failure from `tests/test_utils.py` formatting.

Local verification:

```bash
python -m pytest tests/test_utils.py -q
uvx nox -s lint -- --show-diff-on-failure
```

**Comment by binggao1230** [comment]
Good point. I pushed `b2988d8` to avoid calling `canonicalize_name()` from `is_normalized_name()`.

The implementation now checks the same roundtrip conditions directly: already lowercase, no `_`, no `.`, and no repeated `--`. That keeps the leading/trailing hyphen behavior covered by the new tests, while avoiding the extra `replace()` calls and condensation loop for names that are already normalized.

Local verification:
- `python -m pytest tests/test_utils.py tests/test_requirements.py -q`
- `uvx nox -s lint -- --show-diff-on-failure`

File: src/packaging/utils.py [hunk 1]
```diff
@@ -123,7 +122,12 @@ def is_normalized_name(name: str) -> bool:
     >>> is_normalized_name("Django")
     False
     """
-    return _normalized_regex.fullmatch(name) is not None
+    return (
```
Comment (brettcannon) [inline src/packaging/utils.py]: This is not complete enough as it doesn't stop any non-ASCII characters from being accepted.

**Review by brettcannon — DISMISSED** [review]

File: src/packaging/utils.py (hunk 1 above)
Comment (henryiii) [reply src/packaging/utils.py]: I think it probably is best kept a regex, just a better regex is needed. :)

Comment (binggao1230) [reply src/packaging/utils.py]: Good call — kept it a regex and tightened the character set. `04edeff` restores a module-level `_normalized_regex`:

```python
_normalized_regex = re.compile(r"-?[a-z0-9]+(?:-[a-z0-9]+)*-?", re.ASCII)
```

This rejects non-ASCII (and other non-name characters like `_`, `.`, uppercase, and hyphen runs), while still accepting the leading/trailing single hyphens that `canonicalize_name` leaves in place — so the roundtrip edge cases this PR added stay covered. Added tests for `café`, `naïve-pkg`, `π`, `foo~bar`, etc.

File: src/packaging/utils.py [hunk 2]
```diff
@@ -61,7 +61,11 @@ class InvalidSdistFilename(ValueError):
 _validate_regex = re.compile(
     r"[a-z0-9]|[a-z0-9][a-z0-9._-]*[a-z0-9]", re.IGNORECASE | re.ASCII
 )
-_normalized_regex = re.compile(r"[a-z0-9]+(?:-[a-z0-9]+)*", re.ASCII)
+# A canonicalize_name fixed point: lowercase ASCII alphanumerics joined by single
+# hyphens. canonicalize_name leaves leading/trailing hyphens untouched, so they are
+# accepted here too, but runs of hyphens and any other character (uppercase, "_",
+# ".", or non-ASCII) prevent a name from being already normalized.
```
Comment (henryiii) [inline src/packaging/utils.py]:
```suggestion
# Lowercase ASCII alphanumerics joined by single hyphens. canonicalize_name
# leaves leading/trailing hyphens untouched, so they are accepted here too.
```

**Review by henryiii — COMMENTED** [review]
Though this worries me: the standard clearly indicate the current regex is correct, leading and trailing hyphens are not allowed - that's from https://peps.python.org/pep-0508/#names. The normalization spec was a different PEP: https://peps.python.org/pep-0503/#normalized-names and it does allow leading/trailing hyphens, though I think it's more likely that's an oversight. There are now consolidated into a single place:

https://packaging.python.org/en/latest/specifications/name-normalization (edited to be the correct link now!)

Where you can see the "contradicting" rules next to teach other. Packaging currently implements exactly those rules, though.

(And, in practice, it means you can't put any name in to normalize_name and get a normalized version - leading/trailing symbols break, though the stated guarantee that these are related is currently wrong)

**Comment by binggao1230** [comment]
Good point — you're right, and I think it reframes this PR.

Since PEP 508 names must start and end with an alphanumeric, `-foo`/`foo-` aren't valid names to begin with. Over the domain of *valid* names, `canonicalize_name` never yields a leading/trailing hyphen, so the original `[a-z0-9]+(?:-[a-z0-9]+)*` (with `re.ASCII`) already matches exactly the set of `canonicalize_name` fixed points — and it rejects non-ASCII for free. The edge cases I added only differ for inputs that aren't valid names, where the roundtrip "guarantee" is moot anyway.

So I'd rather drop the behavior change: keep the original regex, and instead just fix the docstring, which currently states the roundtrip relationship without noting it only holds for valid names. Happy to reduce this PR to that (revert the regex + the edge-hyphen test, tighten the docstring), or to close it if you'd prefer to touch the docstring separately. Which would you like?

**Comment by henryiii** [comment]
I think fixing the docstring is the right path forward for now. We should look into changing the standards first if we do want the round trip property.

**Comment by binggao1230** [comment]
Done — reduced to the docstring change. `2d7c2fa` rebases onto current `main` and drops the regex change and the tests I'd added, so `is_normalized_name` behaviour is untouched. The docstring now notes the roundtrip only characterizes normalized names for *valid* names, with doctests for the leading-hyphen case:

```python
>>> canonicalize_name("_not_legal")
'-not-legal'
>>> is_normalized_name("-not-legal")  # roundtrips, but not a valid name
False
```

Diff is `src/packaging/utils.py` only (+12/-3); doctests and ruff pass.

**Comment by brettcannon** [comment]
> Though this worries me: the standard clearly indicate the current regex is correct, leading and trailing hyphens are not allowed - that's from https://peps.python.org/pep-0508/#names. The normalization spec was a different PEP: https://peps.python.org/pep-0503/#normalized-names and it does allow leading/trailing hyphens, though I think it's more likely that's an oversight. There are now consolidated into a single place:
> 
> https://peps.python.org/pep-0508/#names
> 
> Where you can see the "contradicting" rules next to teach other. Packaging currently implements exactly those rules, though.
> 
> (And, in practice, it means you can't put any name in to normalize_name and get a normalized version - leading/trailing symbols break, though the stated guarantee that these are related is currently wrong)

What does packaging.python.org say, as that's the official spec.

**Comment by henryiii** [comment]
Ahh, sorry, I copied and pasted incorrectly. It was supposed to be:

> There are now consolidated into a single place:
> 
> https://packaging.python.org/en/latest/specifications/name-normalization

Instead I linked to https://peps.python.org/pep-0508/#names again, sorry! They sit next to each other there.
